Event size validation for Wodles

RamosFe opened the pull request:
|Related issue|
|---|
| [#17689](https://github.com/wazuh/wazuh/issues/17689) |

<!--
This template reflects sections that must be included in new Pull requests.
Contributions from the community are really appreciated. If this is the case, please add the
"contribution" to properly track the Pull Request.

Please fill the table above. Feel free to extend it at your convenience.
-->

## Description

<!--
Add a clear description of how the problem has been solved.
-->

Added validations for the size of the events sent to the Analysisd queue.

fdalmaup commented on wodles/aws/aws_s3.py:
```diff
@@ -513,8 +511,15 @@ def send_msg(self, msg, dump_json=True):
+
+            # Logs warning if event is bigger than max size
+            if len(encoded_msg) > MAX_EVENT_SIZE:
+                debug(f"WARNING: : Event size exceeds the maximum allowed limit of {MAX_EVENT_SIZE} bytes.")
```
The `debug` function requires the `msg_level` argument

fdalmaup reviewed (changes requested):
The modifications seem right but there is no manual testing evidence in the PR nor the issue. Please add it for at least one of the integrations.

RamosFe commented:
## Update

Created a new file with a size bigger than the limit set in the new validations:

<img width="462" alt="image" src="https://github.com/wazuh/wazuh/assets/37565679/04d71138-6e30-40fa-bf4f-af795a1d58d3">

This file has only 1 line with a string bigger than the max limit (Because the Azure Storage Wodle sends the bia socket file line by line, the file only has 1 line to send the string with the desired size and raise the warning).
Script execution:

```
# ./azure-logs --storage --storage_auth_path /var/ossec/wodles/azure/azure_credentials --container bigblob --reparse
2023/07/21 01:51:46 azure: WARNING: WARNING: Event size exceeds the maximum allowed limit of 65535 bytes.
2023/07/21 01:51:46 azure: ERROR: ERROR: Message too long to send to Wazuh.  Skipping message...
2023/07/21 01:51:47 azure: WARNING: WARNING: Event size exceeds the maximum allowed limit of 65535 bytes.
2023/07/21 01:51:47 azure: ERROR: ERROR: Message too long to send to Wazuh.  Skipping message...
2023/07/21 01:51:47 azure: WARNING: WARNING: Event size exceeds the maximum allowed limit of 65535 bytes.
```

fdalmaup commented on wodles/aws/aws_s3.py:
```diff
@@ -513,8 +511,15 @@ def send_msg(self, msg, dump_json=True):
+                                                 msg=json_msg if dump_json else msg).encode()
+
+            # Logs warning if event is bigger than max size
+            if len(encoded_msg) > MAX_EVENT_SIZE:
```
The AWS module imports the entire utils file. Therefore, this line will throw an error.
```suggestion
            if len(encoded_msg) > utils.MAX_EVENT_SIZE:
```

RamosFe replied:
Fixed in commit 113469afbaff8582af4db430587402c60c697277:

https://github.com/wazuh/wazuh/blob/113469afbaff8582af4db430587402c60c697277/wodles/aws/aws_s3.py#L519-L520

fdalmaup commented on wodles/azure/azure-logs.py:
```diff
@@ -828,9 +828,16 @@ def send_message(message: str):
+
+    # Logs warning if event is bigger than max size
+    if len(encoded_msg) > MAX_EVENT_SIZE:
+        logging.warning(f"WARNING: Event size exceeds the maximum allowed limit of {MAX_EVENT_SIZE} bytes.")
```
You may remove `WARNING` from the log message since the logger already adds it.

RamosFe replied:
Fixed in commit 113469afbaff8582af4db430587402c60c697277:

https://github.com/wazuh/wazuh/blob/113469afbaff8582af4db430587402c60c697277/wodles/aws/aws_s3.py#L519-L520

fdalmaup reviewed (approved):
LGTM!

Selutario reviewed (commented):
In the update https://github.com/wazuh/wazuh/pull/17932#issuecomment-1645635392 I see these two errors:
```
2023/07/21 01:51:46 azure: WARNING: WARNING: Event size exceeds the maximum allowed limit of 65535 bytes.
2023/07/21 01:51:46 azure: ERROR: ERROR: Message too long to send to Wazuh.  Skipping message...
```

The first one is the new warning you added and the second one is the error shown when the response code is `90`:
```
elif e.errno == 90:
    logging.error("ERROR: Message too long to send to Wazuh.  Skipping message...")
``` 

However, according to the [issue description](https://github.com/wazuh/wazuh/issues/17689#issue-1779763054), this error is only supposed to appear when the buffer size of `/var/ossec/queue/sockets/queue` is exceeded, which is 212992B. 

Since the file you uploaded is 93 KiB, far from the size of the buffer, could you verify from what event size the response code 90 is being obtained? If it is returned for events of 65 KiB or bigger, this development would not be necessary.

RamosFe commented:
Added extra error messages only for debugging this case:
### Debugging
```
    try:
        s.connect(ANALYSISD)
        logging.warning(f"WARNING: Event size {len(encoded_msg)}")
        s.send(encoded_msg)
    except for socket_error as e:
        if e.errno == 111:
            logging.error("ERROR: Wazuh must be running.")
            sys.exit(1)
        elif e.errno == 90:
            logging.error("ERROR: Message too long to send to Wazuh.  Skipping message...")
            send_buffer_size = s.getsockopt(SOL_SOCKET, SO_SNDBUF)
            recv_buffer_size = s.getsockopt(SOL_SOCKET, SO_RCVBUF)
            logging.error(f"ERROR: Wazuh send {send_buffer_size} size and recv {recv_buffer_size}")
            logging.error(e)
        else:
            logging.error(f"ERROR: Error sending message to wazuh: {e}")
            sys.exit(1)
    finally:
        s.close()
```
### Result
There are 3 files in the `bigblob` container with the following sizes: 500mb, 5mb, and 93KB. For the first 2 files, the WARNING and ERROR are logged, but for the last file, only the WARNING is logged.
I also added a log message for getting the size of the `send` and `recv` buffers. These two values are always 212992B.

```
./azure-logs --storage --storage_auth_path /var/ossec/wodles/azure/azure_credentials --container bigblob --reparse
2023/08/16 06:44:40 azure: WARNING: WARNING: Event size exceeds the maximum allowed limit of 65535 bytes.
2023/08/16 06:44:40 azure: WARNING: WARNING: Event size 524288034
2023/08/16 06:44:40 azure: ERROR: ERROR: Message too long to send to Wazuh.  Skipping message...
2023/08/16 06:44:40 azure: ERROR: ERROR: Wazuh send 212992 size and recv 212992
2023/08/16 06:44:40 azure: ERROR: [Errno 90] Message too long
2023/08/16 06:44:41 azure: WARNING: WARNING: Event size exceeds the maximum allowed limit of 65535 bytes.
2023/08/16 06:44:41 azure: WARNING: WARNING: Event size 5242914
2023/08/16 06:44:41 azure: ERROR: ERROR: Message too long to send to Wazuh.  Skipping message...
2023/08/16 06:44:41 azure: ERROR: ERROR: Wazuh send 212992 size and recv 212992
2023/08/16 06:44:41 azure: ERROR: [Errno 90] Message too long
2023/08/16 06:44:41 azure: WARNING: WARNING: Event size exceeds the maximum allowed limit of 65535 bytes.
2023/08/16 06:44:41 azure: WARNING: WARNING: Event size 95569
```

Selutario reviewed (changes requested):
Fix conflict in aws wodle: 
![image](https://github.com/wazuh/wazuh/assets/23361101/c6423b46-719b-41c7-a451-cfe4166ae443)

RamosFe commented on wodles/aws/aws_s3.py:
```diff
@@ -540,8 +540,20 @@ def send_msg(self, msg, dump_json=True):
+                debug(f"Event size exceeds the maximum allowed limit of {utils.MAX_EVENT_SIZE} bytes.", 1)
+
+            s.send(encoded_msg)
+>>>>>>> 6831ecae28 (Added msg_level in aws_s3.py.)
```
Fixed in e7a86830c7259c041e2ad24e904b2f2db717851e.
